File: Python/ByTier/C/2123_C_Prefix_Min_and_Suffix_Max.py
# Problem: CF 2123 C - Prefix Min and Suffix Max
# https://codeforces.com/contest/2123/problem/C

"""
C. Prefix Min and Suffix Max
Algorithms/Techniques: Prefix minimum, suffix maximum, greedy approach

Time Complexity: O(n) per test case
Space Complexity: O(n) per test case

For each element in the array, we determine if it's possible to reduce the entire array to just that element using operations:
- Replace a prefix with its minimum value
- Replace a suffix with its maximum value

The key insight is:
1. An element can be the final remaining element only if it is either:
   - The minimum in some prefix (from left)
   - The maximum in some suffix (from right)

2. For each position i, we check:
   - If a[i] is the minimum in any prefix ending at or before i
   - If a[i] is the maximum in any suffix starting at or after i

This can be done efficiently by precomputing:
- Prefix minimums from left to right
- Suffix maximums from right to left

The solution uses two passes:
1. Left-to-right pass to mark elements that could be the final element due to being a prefix minimum
2. Right-to-left pass to mark elements that could be the final element due to being a suffix maximum
"""

import sys

input = sys.stdin.readline
t = int(input())
for _ in range(t):
    n = int(input())
    a = list(map(int, input().split()))
    ans = ["0"] * n
    ans[0] = "1"  # First element can always be the answer (prefix of size 1)
    ans[-1] = "1"  # Last element can always be the answer (suffix of size 1)
    
    mi = a[0]  # Initialize minimum from left
    for i in range(1, n):
        if a[i] < mi:
            ans[i] = "1"
            mi = a[i]
    
    ma = a[-1]  # Initialize maximum from right
    for i in range(n - 2, -1, -1):
        if a[i] > ma:
            ans[i] = "1"
            ma = a[i]
    
    print("".join(ans))


# https://github.com/VaHiX/codeForces/

File: Kotlin/ByTier/C/2123_C_Prefix_Min_and_Suffix_Max.kt
// Problem: CF 2123 C - Prefix Min and Suffix Max
// https://codeforces.com/contest/2123/problem/C

/*
 * Problem: C. Prefix Min and Suffix Max
 * Purpose: For each element in the array, determine if it's possible to reduce the array to a single element equal to that value using operations:
 *          - Replace any prefix with its minimum value
 *          - Replace any suffix with its maximum value
 *
 * Algorithm:
 * 1. For each position i, check if a[i] can be the final remaining element.
 *    This happens if and only if:
 *    - a[i] is the minimum of all elements to its left (including itself)
 *    - a[i] is the maximum of all elements to its right (including itself)
 *
 * Time Complexity: O(n) per test case
 * Space Complexity: O(n) for the result array
 */
fun Solve(a: IntArray): String {
    val n = a.size
    val vecA = CharArray(n) { '0' }
    
    // Mark positions that can be the minimum of some prefix (from left to right)
    var minLeft = a[0]
    vecA[0] = '1'  // First element can always be the minimum of prefix [0..0]
    for (i in 1 until n) {
        if (a[i] < minLeft) {
            minLeft = a[i]
            vecA[i] = '1'
        }
    }
    
    // Mark positions that can be the maximum of some suffix (from right to left)
    var maxRight = a[n - 1]
    vecA[n - 1] = '1'  // Last element can always be the maximum of suffix [n-1..n-1]
    for (i in n - 2 downTo 0) {
        if (a[i] > maxRight) {
            maxRight = a[i]
            vecA[i] = '1'
        }
    }
    
    return vecA.joinToString("")
}

fun main() {
    val t = readLine()!!.toInt()
    repeat(t) {
        val n = readLine()!!.toInt()
        val a = readLine()!!.split(" ").map { it.toInt() }.toIntArray()
        println(Solve(a))
    }
}


// https://github.com/VaHiX/CodeForces/